### ISSUE TITLE
Problems of inconsistent preprocessing and returned data when using PIL and torch images for prediction

### ISSUE THREAD

**Issue by petercham** [ultralytics/ultralytics#16909]
### Search before asking

- [X] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

1.Using torch images in the shape of bchw:
Inconvenience: The imgsz parameter is ineffective. One needs to resize the image to many different sizes on their own to handle images or videos uploaded by users with greatly varying sizes or compositions.
Advantage: The prediction results will organize the results list according to the batch of images. In each result in the list, a unique mask of shape (n, h, w) or box of shape (n, 4) is generated according to the number of detected objects.

2.Using PIL images:
Advantage: The imgsz parameter can be used without the need to write one's own image preprocessing. However, imgsz has a relatively large impact on the prediction results, and it is necessary to design and optimize the prediction size algorithm to ensure that an imgsz with the expected confidence can be found. This can be flexibly programmed by oneself.
Inconvenience: The predicted results will organize the results list according to a single image. The n results in the list represent the n detected objects. Each data is returned in the shape of a mask of (1, h, w) or a box of (1, 4). Therefore, one needs to modify the result data formatting method on their own. It would be great if the results could be the same as those of torch, generating a mask of shape (n, h, w) or a box of shape (n, 4). Then a single set of data formatting methods can be universally used in different scenarios.

3.Both of the above methods return data according to the final actual preprocessed image size. Therefore, one must manually restore the width and height dimensions of the prediction result data to the original image size. It would be excellent if the data with the restored original image size could be directly returned.

### Use case

1. When inputting a torch.Tensor image in the shape of bchw, it should support the convenient use of `imgsz` for image preprocessing without the need to write additional preprocessing code on one's own.
2. When inputting a PIL image, the returned results should be the same as those for torch images, returned in the shapes of (n, h, w) and (n, 4).
3. It is hoped that the returned result data can be restored to the original size of the input image.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @petercham, thank you for reaching out to Ultralytics 🚀! We understand the challenges you're facing with the inconsistencies in preprocessing and data returns when using PIL and torch images.

For a 🐛 Bug Report, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/)? This would help us investigate further.

In the meantime, we recommend checking our [Docs](https://docs.ultralytics.com) for more details on image processing and [model training tips](https://docs.ultralytics.com/guides/model-training-tips/). You might find useful examples for overcoming these inconsistencies.

Feel free to join our community for more interactive support:
- Jump into our [Discord](https://ultralytics.com/discord) 🎧 for real-time chat
- Engage in deeper discussions on [Discourse](https://community.ultralytics.com)
- Join our [Subreddit](https://reddit.com/r/ultralytics) for community knowledge sharing

## Upgrade

To ensure you’re encountering the issue with the latest updates, upgrade to the latest `ultralytics` version and ensure all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are met in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs smoothly in various environments. Choose one that suits your needs:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: Deep Learning VM [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: Deep Learning AMI [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

Check the Ultralytics CI below to verify the operational status across different platforms:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This is an automated response, but rest assured, an Ultralytics engineer will review your issue soon! 😊

**Comment by Y-T-G** [comment]
1. Why not convert to `numpy` array? The `tensor` images are supposed to be preprocessed.

3. `result.plot()` provides the annotated image in the original shape. `result.orig_img` provides the original unaltered image.

**Comment by petercham** [comment]
> 1. Why not convert to `numpy` array? The `tensor` images are supposed to be preprocessed.
> 2. `result.plot()` provides the annotated image in the original shape. `result.orig_img` provides the original unaltered image.

Because my application scenario is mainly in the image analysis stage of TensorFlow. Most of the time, the image is torch in the shape of bhwc. The data structure returned by YOLO itself is also torch. In fact, I hope to minimize the conversion of object types in the middle. So I hope that the input of torch.Tensor can also support completing preprocessing through imgsz.

The purpose of restoring the data shape is to implement subsequent operations such as cutout or fusion. So plotting only serves as a debugging function. Before applying the data, manual restoration is still needed.

**Comment by Y-T-G** [comment]
> So I hope that the input of torch.Tensor can also support completing preprocessing through imgsz.

The preprocessing functions require `numpy` arrays since they use `cv2`, so it will have to be converted to `numpy` whether you do it manually or `ultralytics` does it internally. So might as well convert it yourself. A similar request was made at https://github.com/ultralytics/ultralytics/issues/16677


> The purpose of restoring the data shape is to implement subsequent operations such as cutout or fusion. So plotting only serves as a debugging function. Before applying the data, manual restoration is still needed.

You can access the tensor using `result.boxes.data` or `result.masks.data`. And the results are scaled with respect to the original image size, so I'm not sure why you think:
>Both of the above methods return data according to the final actual preprocessed image size.

**Comment by petercham** [comment]
> > So I hope that the input of torch.Tensor can also support completing preprocessing through imgsz.
> 
> The preprocessing functions require `numpy` arrays since they use `cv2`, so it will have to be converted to `numpy` whether you do it manually or `ultralytics` does it internally. So might as well convert it yourself. A similar request was made at #16677
> 
> > The purpose of restoring the data shape is to implement subsequent operations such as cutout or fusion. So plotting only serves as a debugging function. Before applying the data, manual restoration is still needed.
> 
> You can access the tensor using `result.boxes.data` or `result.masks.data`. And the results are scaled with respect to the original image size, so I'm not sure why you think:
> 
> > Both of the above methods return data according to the final actual preprocessed image size.

I am using the data extracted by `boxes.xyxy` and `masks.data`.
For example, for a mask: `[{'data': tensor([[[0., 0., 0.,..., 0., 0., 0.], [0., 0., 0.,..., 0., 0., 0.], [0., 0., 0.,..., 0., 0., 0.],..., [0., 0., 0.,..., 0., 0., 0.], [0., 0., 0.,..., 0., 0., 0.], [0., 0., 0.,..., 0., 0., 0.]]]), 'orig_shape': (2160, 3840), 'shape': torch.Size([1, 160, 256]), 'xy': [array([[...]]`. This is the dictionary I converted from the original returned data. When applying it, I found that the shape of the mask is not `(1, 2160, 3840)`, but `torch.Size([1, 160, 256])`. The same goes for `boxes`.
After using my own written code to restore the shape, it is restored to the original width and height of the image:
                # Get the width and height of the original image
                orig_w, orig_h = self.orig_shape
                # Adjust the mask size according to the accurate mode
                if self.accurate:
                    # In high-precision mode, the all-zero area padded needs to be cropped.
                    mask_data = Resize.align_topleft(formatted_result['data'], orig_w, orig_h)
                    logger.info(f"IS Accurate Mode, use Resize.align_topleft for Restore.")
                else:
                    # In performance mode, it needs to be scaled and the all-zero area padded needs to be cropped.
                    mask_data = Resize.crop_topleft(formatted_result['data'], orig_w, orig_h)
                    logger.info(f"Not Accurate Mode, use Resize.fill_topleft for Restore.")
                # Update the 'data' field in formatted_result
                formatted_result['data'] = mask_data.

This is the case whether the input is pil or torch.

**Comment by Y-T-G** [comment]
The boxes are scaled in post-processing to match original image size.

https://github.com/ultralytics/ultralytics/blob/5a9ce863e41ccb685dd1e31f0e94417b1eef5ad3/ultralytics/models/yolo/segment/predict.py#L49-L50

To scale the masks to original image, you can pass `retina_masks=True`
